### PR TITLE
ARROW-5707: [Java] Improve the performance and code structure for ArrowRecordBatch

### PR DESCRIPTION
**Pull request by liyafan82**
Improve the performance of ArrowRecordBatch by reducing the number of divisions.
Improve the code structure of ArrowRecordBatch by removing the useless constructor.

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java [hunk 1]
```diff
@@ -213,9 +197,7 @@ public int computeBodyLength() {
       ByteBuffer nioBuffer =
           buffer.nioBuffer(buffer.readerIndex(), buffer.readableBytes());
       size += nioBuffer.remaining();
-      if (size % 8 != 0) {
-        size += 8 - (size % 8);
-      }
+      size = (size + 7) & 0xfffffff8;
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Please add a comment here.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java [hunk 2]
```diff
@@ -79,29 +79,13 @@ public ArrowRecordBatch(int length, List<ArrowFieldNode> nodes, List<ArrowBuf> b
...
-  // clone constructor
-  private ArrowRecordBatch(boolean dummy, int length, List<ArrowFieldNode> nodes, List<ArrowBuf> buffers) {
-    this.length = length;
-    this.nodes = nodes;
-    this.buffers = buffers;
-    this.closed = false;
-    List<ArrowBuffer> arrowBuffers = new ArrayList<>();
-    long offset = 0;
-    for (ArrowBuf arrowBuf : buffers) {
-      long size = arrowBuf.readableBytes();
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: it looks like this doesn't call   arrowBuf.getReferenceManager().retain();

That seems like a functionality difference?

Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Also this changes removing this causes the buffers to be aligned by default, where they where unaligned before.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java (hunk 1 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Sure. Thanks.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java (hunk 2 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Nice catch.

This constructor is functionally different from the public constructors. That is why it is in the code. 
I have restored this constructor and added comments for it.

**Comment by emkornfield** [comment]
@liyafan82 is there a benchmark showing the performance improvement?

In general, I think we should be careful of over-optimizing code that isn't on the hot path.

**Comment by liyafan82** [comment]
> @liyafan82 is there a benchmark showing the performance improvement?
> 
> In general, I think we should be careful of over-optimizing code that isn't on the hot path.

Sure. I totally agree with you that we should be careful not to over-optimize the code, especially for performance improves that harm code readability.

According to our experience, ArrowRecordBatch is intensively used when doing serialization/deserialization. For example, in the shuffling phase of a distributed computing engine, in the upstream task, ArrowRecordBatch is used to convert vectors to byte streams while in the downstream task, ArrowRecordBatch is used to convert byte stream to vectors, so ArrowRecordBatches are frequently created and the computeBodyLength is frequently called. 

Therefore, ArrowRecordBatch is on the hot path in our scenario.

Having said that, it should be noted that the performance improvement for this particular PR is insignificant:

Before:
ArrowRecordBatchBenchmarks.createAndGetLength  avgt    5  355.399 ± 16.183  ns/op

After:
ArrowRecordBatchBenchmarks.createAndGetLength  avgt    5  352.506 ± 14.307  ns/op

This could possibly means:
1. The integer operations only account for a small part of the overall run time.
2. More improvements may be needed.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java [hunk 3]
```diff
@@ -79,14 +79,17 @@ public ArrowRecordBatch(int length, List<ArrowFieldNode> nodes, List<ArrowBuf> b
       arrowBuffers.add(new ArrowBuffer(offset, size));
       LOGGER.debug("Buffer in RecordBatch at {}, length: {}", offset, size);
       offset += size;
-      if (alignBuffers && offset % 8 != 0) { // align on 8 byte boundaries
-        offset += 8 - (offset % 8);
+      if (alignBuffers) { // align on 8 byte boundaries
+        offset = (offset + 7) & 0xfffffffffffffff8L;
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: there are a few places where this code or similar occurs, can we find a centralized place to put it (same with code below).

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java [hunk 4]
```diff
@@ -79,14 +79,17 @@ public ArrowRecordBatch(int length, List<ArrowFieldNode> nodes, List<ArrowBuf> b
       arrowBuffers.add(new ArrowBuffer(offset, size));
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: if you are looking for a more substantial win, you might try passing in the initial capacity to arrowBuffers above.

**Comment by emkornfield** [comment]
> The integer operations only account for a small part of the overall run time. More improvements may be needed.
I think this is the case which is why we should be careful.  Based on the variance it is not clear if this is actually repeatable (if you run the bencmark a second time do you see the same difference?)

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java (hunk 3 above)
Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Specifically:
vector/src//main/java/org/apache/arrow/vector/BaseValueVector.java:    return (valueCount + 7) >> 3;
vector/src//main/java/org/apache/arrow/vector/BaseValueVector.java:    return ((size + 7) / 8) * 8;
vector/src//main/java/org/apache/arrow/vector/BitVectorHelper.java:    return (valueCount + 7) >> 3;

Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Good suggestion. I have opened a separate class to place such utilities.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java (hunk 4 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowRecordBatch.java]: Thanks for the good suggestion.

**Comment by liyafan82** [comment]
> > The integer operations only account for a small part of the overall run time. More improvements may be needed.
> > I think this is the case which is why we should be careful.  Based on the variance it is not clear if this is actually repeatable (if you run the bencmark a second time do you see the same difference?)

I agree with you that the insignificant improvements make the performance results unstable.
However, JIT analysis results show that the improvements are sure to win. 

For example, for the first improvement of ArrowRecordBatch, the two divisions are unavoidable (although they are optimized as two shift instructions).

image

In contrast, our improved code has only one add instruction.

For the second improvement of ArrowRecordBatch, the branch cannot be optimized away by the JIT:

image

In our improved code, the branch no longer exists.

File: java/vector/src/main/java/org/apache/arrow/util/DataSizeRoundingUtil.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.util;
+
+/**
+ * Utilities for rounding data size.
+ */
+public class DataSizeRoundingUtil {
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/util/DataSizeRoundingUtil.java]: make the class final?

**Comment by emkornfield** [comment]
+1, LGTM, trying to rerun failed jobs.

File: java/vector/src/main/java/org/apache/arrow/util/DataSizeRoundingUtil.java (hunk 5 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/util/DataSizeRoundingUtil.java]: Good suggestion. Thanks.

**Comment by emkornfield** [comment]
Thank you.

**Pull request closed without merging**